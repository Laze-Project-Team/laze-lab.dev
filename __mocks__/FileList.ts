class FileList<T> {
  _files: T[];
  length: number;
  constructor(files: T[]) {
    this._files = files;
    this.length = this._files.length;
  }
  item(index: number) {
    return this._files[index];
  }
  get [Symbol.toStringTag]() {
    return 'FileList';
  }
}

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
window.FileList = FileList;

export {};
