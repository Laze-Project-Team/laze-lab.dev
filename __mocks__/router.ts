import type { useRouter } from 'next/router';
import { beforeEach, vi } from 'vitest';

export const mockRouter = (): void => {
  vi.mock('next/router', () => ({
    useRouter: () => {
      const router: ReturnType<typeof useRouter> = {
        asPath: '/',
        back: vi.fn(),
        basePath: '/',
        beforePopState: vi.fn(),
        events: {
          on: vi.fn(),
          off: vi.fn(),
          emit: vi.fn(),
        },
        isFallback: false,
        isLocaleDomain: false,
        isPreview: false,
        isReady: true,
        locale: 'en',
        locales: ['en'],
        prefetch: vi.fn(),
        push: vi.fn(),
        query: {},
        reload: vi.fn(),
        replace: vi.fn(),
        route: '/',
        pathname: '/',
      };

      return router;
    },
  }));
};

export const mockEachRouter = (): void => {
  beforeEach(() => {
    mockRouter();
  });
};
