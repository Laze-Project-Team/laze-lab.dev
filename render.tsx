import type {
  Queries,
  queries,
  RenderOptions,
  RenderResult,
} from '@testing-library/react';
import { render as defaultRender } from '@testing-library/react';
import type { FC, ReactNode } from 'react';

import { ColorModeProvider } from '@/components/contexts/ColorModeContext';

export type renderType<
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container,
> = (
  ui: React.ReactElement,
  options?: RenderOptions<Q, Container, BaseElement>,
) => RenderResult<Q, Container, BaseElement>;

type wrapperProps = {
  children?: ReactNode;
};

type globalWrapperHOCType = (
  wrapper?: React.JSXElementConstructor<{ children: React.ReactElement }>,
) => FC<wrapperProps>;

const globalWrapperHOC: globalWrapperHOCType = (Wrapper) => {
  const globalWrapper: FC<wrapperProps> = ({ children }) => (
    <ColorModeProvider>
      {Wrapper ? (
        <Wrapper>
          <>{children}</>
        </Wrapper>
      ) : (
        children
      )}
    </ColorModeProvider>
  );

  return globalWrapper;
};

export const render: renderType = (ui, options) =>
  defaultRender(ui, {
    ...options,
    wrapper: globalWrapperHOC(options?.wrapper),
  });
