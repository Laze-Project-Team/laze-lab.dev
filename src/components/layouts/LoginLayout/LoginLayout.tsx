import { css } from '@emotion/react';
import ArrowBackIosIcon from '@mui/icons-material/ArrowBackIos';
import Container from '@mui/material/Container';
import { useTranslation } from 'next-i18next';
import type { FC, ReactNode } from 'react';

import { DefaultLink } from '@/components/ui/DefaultLink';
import { pagesPath } from '@/lib/$path';

export type loginLayoutProps = {
  title: ReactNode;
  children?: ReactNode;
};

export type presentialLoginLayoutProps = loginLayoutProps;

export const PresentialLoginLayout: FC<presentialLoginLayoutProps> = ({
  title,
  children,
}) => {
  const [t] = useTranslation('common');

  return (
    <Container maxWidth="sm">
      <main>
        <h1
          css={css`
            margin: 2rem 0;
            font-size: 1.5rem;
            text-align: center;
          `}
        >
          {title}
        </h1>

        <DefaultLink href={pagesPath.$url()}>
          <ArrowBackIosIcon sx={{ fontSize: '1rem' }} />
          <span>{t('backToHome')}</span>
        </DefaultLink>

        {children}
      </main>
    </Container>
  );
};

export const LoginLayout: FC<loginLayoutProps> = (props) => {
  return <PresentialLoginLayout {...props} />;
};
