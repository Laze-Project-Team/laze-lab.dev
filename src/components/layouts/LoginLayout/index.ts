export type { loginLayoutProps } from './LoginLayout';
export { LoginLayout } from './LoginLayout';
