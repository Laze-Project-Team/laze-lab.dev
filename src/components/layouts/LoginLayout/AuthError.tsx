import { Alert } from '@mantine/core';
import type { FC, ReactNode } from 'react';
import { createContext, useState } from 'react';

import { useAuthError } from '@/components/layouts/LoginLayout/useAuthError';

export type authErrorContextType = {
  error: string | null;
  setError: (error: string | null) => void;
};

export const authErrorContext = createContext<authErrorContextType | null>(
  null,
);

export type authError = {
  code: string;
  message: string;
};

export type authErrorsProviderProps = {
  children: ReactNode;
};

export const AuthErrorProvider: FC<authErrorsProviderProps> = ({
  children,
}) => {
  const [error, setError] = useState<string | null>(null);

  return (
    <authErrorContext.Provider value={{ error, setError }}>
      {children}
    </authErrorContext.Provider>
  );
};

export const AuthError: FC = () => {
  const { error, setError } = useAuthError();

  return <>{error && <Alert onClose={() => setError(null)}>{error}</Alert>}</>;
};
