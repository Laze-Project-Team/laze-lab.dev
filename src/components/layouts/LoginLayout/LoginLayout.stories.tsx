import type { ComponentMeta, ComponentStory } from '@storybook/react';

import { PresentialLoginLayout } from './LoginLayout';

export default {
  title: 'layouts/LoginLayout',
  component: PresentialLoginLayout,
} as ComponentMeta<typeof PresentialLoginLayout>;

const Template: ComponentStory<typeof PresentialLoginLayout> = (props) => (
  <PresentialLoginLayout {...props} />
);

export const Primary = Template.bind({});
Primary.args = {
  title: 'Here is the title',
  children: 'Here is the children',
};
