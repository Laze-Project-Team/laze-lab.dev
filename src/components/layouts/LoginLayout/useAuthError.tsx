import { useContext } from 'react';

import type { authError } from './AuthError';
import { authErrorContext } from './AuthError';

export type authErrorHandlers = {
  error: string | null;
  setError: (errorMessage: string | null) => void;
  handleError: (error: authError) => void;
};

export const useAuthError = (): authErrorHandlers => {
  const authCtx = useContext(authErrorContext);

  if (authCtx === null) {
    throw new Error('Please wrap the component with AuthErrorProvider');
  }

  const { error, setError } = authCtx;

  return {
    error,
    setError: (errorMessage) => {
      setError(errorMessage);
    },
    handleError: (err) => {
      setError(err.message);
    },
  };
};
