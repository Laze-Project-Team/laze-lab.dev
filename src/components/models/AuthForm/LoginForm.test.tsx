import '@testing-library/jest-dom';

import { fireEvent, screen } from '@testing-library/react';
import { expect, it } from 'vitest';

import { mockEachRouter } from '/__mocks__/router';
import { AuthErrorProvider } from '@/components/layouts/LoginLayout/AuthError';
import { render } from '@/lib/test/render';

import { LoginForm } from '.';

mockEachRouter();

it('should be rendered correctly', () => {
  const { container } = render(<LoginForm />, { wrapper: AuthErrorProvider });
  expect(container.firstChild).toMatchSnapshot();
});

it('should have "form" role', () => {
  render(<LoginForm />, { wrapper: AuthErrorProvider });
  expect(screen.getByRole('form')).toBeInTheDocument();
});

it('should work as login form', () => {
  const form = render(<LoginForm />, { wrapper: AuthErrorProvider });

  const emailInput = form.getByRole('textbox', {
    name: 'form.label.email',
  }) as HTMLInputElement;
  const emailValue = 'aaa-bbb-ccc@example.com';
  fireEvent.change(emailInput, {
    target: { value: emailValue },
  });
  expect(emailInput.value).toBe(emailValue);

  const passwordInput = form.getByRole('textbox', {
    name: 'form.label.password',
  }) as HTMLInputElement;
  const passwordValue = 'this is password';
  fireEvent.change(passwordInput, {
    target: { value: passwordValue },
  });
  expect(passwordInput.value).toBe(passwordValue);
});
