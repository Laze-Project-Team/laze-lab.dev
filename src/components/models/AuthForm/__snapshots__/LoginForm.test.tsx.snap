// Vitest Snapshot v1

exports[`should be rendered correctly 1`] = `
<div
  aria-hidden="true"
  class="MuiBackdrop-root css-17lf3el-MuiBackdrop-root"
  style="opacity: 0; visibility: hidden;"
>
  <span
    class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorInherit css-62e83j-MuiCircularProgress-root"
    role="progressbar"
    style="width: 40px; height: 40px;"
  >
    <svg
      class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </span>
</div>
`;
